File: src/mysequator/engine/postprocess.py
from __future__ import annotations

import numpy as np
from PIL import Image, ImageFilter

from .io import luminance


def apply_auto_brightness(image: np.ndarray, target_percentile: float = 99.7, target_value: float = 0.92) -> np.ndarray:
    value = float(np.percentile(luminance(image), target_percentile))
    if value <= 1e-6:
        return image
    scale = target_value / value
    return np.clip(image * scale, 0.0, 1.0)


def apply_hdr_stretch(image: np.ndarray) -> np.ndarray:
    lo = np.percentile(image.reshape(-1, image.shape[-1]), 0.2, axis=0)
    hi = np.percentile(image.reshape(-1, image.shape[-1]), 99.9, axis=0)
    span = np.maximum(hi - lo, 1e-5)
    return np.clip((image - lo.reshape(1, 1, -1)) / span.reshape(1, 1, -1), 0.0, 1.0)


def reduce_light_pollution(image: np.ndarray, strength: float = 0.45) -> np.ndarray:
    """Subtract a broad low-frequency color background estimate."""
    h, w = image.shape[:2]
    small_w = max(16, min(96, w // 24))
    small_h = max(16, min(96, h // 24))
    background = np.zeros_like(image)

    for channel in range(3):
        channel_u16 = np.round(np.clip(image[:, :, channel], 0.0, 1.0) * 65535).astype(np.uint16)
        pil = Image.fromarray(channel_u16, mode="I;16")
        low = pil.resize((small_w, small_h), resample=Image.BILINEAR)
        low = low.filter(ImageFilter.GaussianBlur(radius=2))
        high = low.resize((w, h), resample=Image.BICUBIC)
        background[:, :, channel] = np.asarray(high, dtype=np.float32) / 65535.0

    neutral = np.percentile(background.reshape(-1, 3), 5, axis=0).reshape(1, 1, 3)
    corrected = image - strength * (background - neutral)
    return np.clip(corrected, 0.0, 1.0)


def enhance_stars(image: np.ndarray, strength: float = 0.35) -> np.ndarray:
    """Mild unsharp-mask enhancement aimed at small bright details."""
    h, w = image.shape[:2]
    enhanced = np.empty_like(image)
    for channel in range(3):
        channel_u16 = np.round(np.clip(image[:, :, channel], 0.0, 1.0) * 65535).astype(np.uint16)
        pil = Image.fromarray(channel_u16, mode="I;16")
        blur = pil.filter(ImageFilter.GaussianBlur(radius=max(1.0, min(h, w) / 900.0)))
        base = np.asarray(pil, dtype=np.float32) / 65535.0
        smooth = np.asarray(blur, dtype=np.float32) / 65535.0
        detail = np.maximum(base - smooth, 0.0)
        enhanced[:, :, channel] = base + strength * detail
    return np.clip(enhanced, 0.0, 1.0)

File: Sources/MySequatorCore/Postprocess.swift
import Foundation

func applyAutoBrightness(_ image: FloatRGBImage, targetPercentile: Float = 99.7, targetValue: Float = 0.92) -> FloatRGBImage {
    let value = image.luminance().percentile(targetPercentile)
    guard value > 1e-6 else { return image }
    let scale = targetValue / value
    var out = image
    for i in out.data.indices {
        out.data[i] = min(max(out.data[i] * scale, 0), 1)
    }
    return out
}

func applyHDRStretch(_ image: FloatRGBImage) -> FloatRGBImage {
    var lows = [Float](repeating: 0, count: 3)
    var highs = [Float](repeating: 1, count: 3)
    for channel in 0..<3 {
        var values = [Float]()
        values.reserveCapacity(image.pixelCount)
        for pixel in 0..<image.pixelCount {
            values.append(image.data[pixel * 3 + channel])
        }
        lows[channel] = values.percentile(0.2)
        highs[channel] = values.percentile(99.9)
    }

    var out = image
    for pixel in 0..<image.pixelCount {
        for channel in 0..<3 {
            let index = pixel * 3 + channel
            let span = max(highs[channel] - lows[channel], 1e-5)
            out.data[index] = min(max((out.data[index] - lows[channel]) / span, 0), 1)
        }
    }
    return out
}

func reduceLightPollution(_ image: FloatRGBImage, strength: Float = 0.45) -> FloatRGBImage {
    let smallWidth = max(16, min(96, image.width / 24))
    let smallHeight = max(16, min(96, image.height / 24))
    var background = FloatRGBImage(width: image.width, height: image.height)

    for channel in 0..<3 {
        var channelData = Array(repeating: Float(0), count: image.pixelCount)
        for pixel in 0..<image.pixelCount {
            channelData[pixel] = image.data[pixel * 3 + channel]
        }
        let low = resizeChannel(channelData, width: image.width, height: image.height, newWidth: smallWidth, newHeight: smallHeight)
        let blurred = blurChannel(low, width: smallWidth, height: smallHeight, radius: 2)
        let high = resizeChannel(blurred, width: smallWidth, height: smallHeight, newWidth: image.width, newHeight: image.height)
        for pixel in 0..<image.pixelCount {
            background.data[pixel * 3 + channel] = high[pixel]
        }
    }

    var neutral = [Float](repeating: 0, count: 3)
    for channel in 0..<3 {
        var values = [Float]()
        values.reserveCapacity(image.pixelCount)
        for pixel in 0..<image.pixelCount {
            values.append(background.data[pixel * 3 + channel])
        }
        neutral[channel] = values.percentile(5)
    }

    var out = image
    for pixel in 0..<image.pixelCount {
        for channel in 0..<3 {
            let index = pixel * 3 + channel
            out.data[index] = min(max(image.data[index] - strength * (background.data[index] - neutral[channel]), 0), 1)
        }
    }
    return out
}

func enhanceStars(_ image: FloatRGBImage, strength: Float = 0.35) -> FloatRGBImage {
    var out = image
    for channel in 0..<3 {
        var channelData = Array(repeating: Float(0), count: image.pixelCount)
        for pixel in 0..<image.pixelCount {
            channelData[pixel] = image.data[pixel * 3 + channel]
        }
        let blurred = blurChannel(channelData, width: image.width, height: image.height, radius: max(1, min(image.width, image.height) / 900))
        for pixel in 0..<image.pixelCount {
            let index = pixel * 3 + channel
            let detail = max(image.data[index] - blurred[pixel], 0)
            out.data[index] = min(max(image.data[index] + strength * detail, 0), 1)
        }
    }
    return out
}

private func resizeChannel(_ image: [Float], width: Int, height: Int, newWidth: Int, newHeight: Int) -> [Float] {
    if width == newWidth, height == newHeight {
        return image
    }
    var out = Array(repeating: Float(0), count: newWidth * newHeight)
    for y in 0..<newHeight {
        let sourceY = Float(y) * Float(height - 1) / Float(max(newHeight - 1, 1))
        let y0 = Int(sourceY.rounded(.down))
        let y1 = min(y0 + 1, height - 1)
        let ty = sourceY - Float(y0)
        for x in 0..<newWidth {
            let sourceX = Float(x) * Float(width - 1) / Float(max(newWidth - 1, 1))
            let x0 = Int(sourceX.rounded(.down))
            let x1 = min(x0 + 1, width - 1)
            let tx = sourceX - Float(x0)
            let a = image[y0 * width + x0] * (1 - tx) + image[y0 * width + x1] * tx
            let b = image[y1 * width + x0] * (1 - tx) + image[y1 * width + x1] * tx
            out[y * newWidth + x] = a * (1 - ty) + b * ty
        }
    }
    return out
}

private func blurChannel(_ image: [Float], width: Int, height: Int, radius: Int) -> [Float] {
    let r = max(radius, 1)
    var temp = Array(repeating: Float(0), count: image.count)
    var out = Array(repeating: Float(0), count: image.count)

    for y in 0..<height {
        for x in 0..<width {
            var sum: Float = 0
            var count: Float = 0
            for xx in max(0, x - r)...min(width - 1, x + r) {
                sum += image[y * width + xx]
                count += 1
            }
            temp[y * width + x] = sum / count
        }
    }

    for y in 0..<height {
        for x in 0..<width {
            var sum: Float = 0
            var count: Float = 0
            for yy in max(0, y - r)...min(height - 1, y + r) {
                sum += temp[yy * width + x]
                count += 1
            }
            out[y * width + x] = sum / count
        }
    }
    return out
}
